src/liveshare/share.ts

```diff
@@ -0,0 +1,310 @@
+import { Commands, Callback, onRequest, request } from './shareCommands';
+import { HelpFile } from '../rHelp';
+
+/// LiveShare
+export let rHostService: HostService = undefined;
+export let rGuestService: GuestService = undefined;
+export let liveSession: vsls.LiveShare;
+export let isGuestSession: boolean;
+export let _sessionStatusBarItem: vscode.StatusBarItem;
+const disposables: vscode.Disposable[] = [];
+
+// service vars
+export const ShareProviderName = 'vscode-r';
+export let service: vsls.SharedServiceProxy | vsls.SharedService | null = undefined;
+
+// random number to fake a UUID for differentiating between
+// host calls and guest calls (specifically for the workspace
+// viewer 'View' function)
+export const UUID = Math.floor(Math.random() * Date.now());
+
+/// state-tracking bools
+// Bool to check if live share is loaded and active
+export function isLiveShare(): boolean {
+    const shareStarted = liveSession?.session?.id;
+    // If there is a hosted session*, return true
+    // else return false
+    // * using vsls.getApi() instead of vsls.getApi().session.id
+    // * will always return true, even if a session is not active
+    // * (a session id will only exist if a session is active)
+    return !!shareStarted;
+}
+
+export function isGuest(): boolean {
+    if (isLiveShare() === true) {
+        return liveSession.session.role === vsls.Role.Guest;
+    } else {
+        return false;
+    }
+}
+
+export function isHost(): boolean {
+    if (isLiveShare() === true) {
+        return liveSession.session.role === vsls.Role.Host;
+    } else {
+        return false;
+    }
+}
+
+// Initialises the Liveshare functionality for host & guest
+// * session watcher is required *
+export async function initLiveShare(context: vscode.ExtensionContext): Promise<void> {
+    if (enableSessionWatcher) {
+        await LiveSessionListener();
+        isGuestSession = isGuest();
+        if (!isGuestSession) {
+            // Construct tree view for host
+            initTreeView();
+        } else {
+            // Construct guest session watcher
+            initGuest(context);
+        }
+
+        // Set context value for hiding buttons for guests
+        void vscode.commands.executeCommand('setContext', 'r.liveShare:isGuest', isGuestSession);
+
+        // push commands
+        context.subscriptions.push(
+            vscode.commands.registerCommand('r.attachActiveGuest', () => attachActiveGuest()),
+            vscode.commands.registerCommand(
+                'r.liveShare.toggle', (node: ToggleNode) => node.toggle(rLiveShareProvider)
+            )
+        );
+    }
+}
+
+// Listens for the activation of a LiveShare session
+export async function LiveSessionListener(): Promise<void> {
+    rHostService = new HostService;
+    rGuestService = new GuestService;
+
+    // Return out when the vsls extension isn't
+    // installed/available
+    const liveSessionStatus = await vsls.getApi();
```

Comment:
When I enabled LiveShare extension on macOS and start debugging the extension or have the extension built and installed, vscode-R could not be fully activated but still stuck here.

Reply:
Is it possible to use a timeout here?

Reply:
What about

```r
const liveSessionStatus = await Promise.race([
        vsls.getApi(),
        new Promise((res) => setTimeout(() => res(null), 1000))
    ]);
    if (!liveSessionStatus) { return; }
    liveSession = liveSessionStatus as vsls.LiveShare;
```

Reply:
Does the code in vscode-jupyter help? https://github.com/microsoft/vscode-jupyter/blob/b1f70d0fc436c9bade9941406b2590b4034cf2bf/src/client/datascience/liveshare/liveshare.ts#L47-L61

Reply:
I like the timeout solution you wrote above - it works nicely with a setting to tweak the timeout speed.

We could, like vscode-jupyter, have a bool to hard-toggle LiveShare functionality. I wonder if it might get confusing with overlap in functionality with the PR's tree view commands?